src/rust/engine/fs/src/store.rs

```diff
@@ -0,0 +1,238 @@
+use bazel_protos;
+use digest::{Digest, FixedOutput};
+use lmdb_rs::{DbFlags, DbHandle, EnvBuilder, Environment, MdbError, MdbValue, ToMdbValue};
+use protobuf::core::Message;
+use sha2::Sha256;
+use std::error::Error;
+use std::mem::transmute;
+use std::path::Path;
+
+use hash::Fingerprint;
+
+///
+/// A content-addressed store of file contents, and Directories.
+///
+/// Currently, Store only stores things locally on disk, but in the future it will gain the ability
+/// to fetch files from remote content-addressable storage too.
+///
+pub struct Store {
+  env: Environment,
+  file_store: DbHandle,
+
+  // Store directories separately from files because:
+  //  1. They may have different lifetimes.
+  //  2. It's nice to know whether we should be able to parse something as a proto.
+  directory_store: DbHandle,
+}
+
+impl Store {
+  pub fn new<P: AsRef<Path>>(path: P) -> Result<Store, String> {
+    // 2 DBs; one for file contents, one for directories.
+    let env = EnvBuilder::new().max_dbs(2).open(path, 0o600).map_err(
+      |e| {
+        format!("Error making env: {}", e.description())
+      },
+    )?;
+    let file_db_handle = env.create_db("files", DbFlags::empty()).map_err(|e| {
+      format!("Error creating/opening files database: {}", e.description())
+    })?;
+    let directory_db_handle = env.create_db("directories", DbFlags::empty()).map_err(
+      |e| {
+        format!(
+          "Error creating/opening directories database: {}",
+          e.description()
+        )
+      },
+    )?;
+    Ok(Store {
+      env: env,
+      file_store: file_db_handle,
+      directory_store: directory_db_handle,
+    })
+  }
+
+  pub fn store_file_bytes(&self, bytes: &[u8]) -> Result<Fingerprint, String> {
+    self.store_bytes(bytes, &self.file_store)
+  }
+
+  fn store_bytes(&self, bytes: &[u8], db: &DbHandle) -> Result<Fingerprint, String> {
+    let mut hasher = Sha256::default();
+    hasher.input(&bytes);
+    let fingerprint = Fingerprint::from_bytes_unsafe(hasher.fixed_result().as_slice());
```

Comment:
**Important**: As it stands, `Fingerprints` are blake2. But since we're going to be forced down the Sha256 path, maybe you can just drop that dependency here and switch these all to Sha256.

Reply:
I'll send a separate PR which this can sit on top of which does that.

Do I need to worry about any kind of migration, or are we happy to assume the two will never collide, and waste whatever disk space we're currently wasting with never-used-again files?

Reply:
Sent https://github.com/pantsbuild/pants/pull/5014 - can discuss over there